core/src/main/java/org/kohsuke/stapler/RequestImpl.java

```diff
@@ -698,20 +698,35 @@ public Object convertJSON(Object o) {
                 if(l==null) {// single value conversion
                     try {
                         Class actualType = type;
+                        boolean isArray = false;
                         String className = null;
                         if(j.has("stapler-class")) {
+                            if (j.optJSONArray("stapler-class") != null) {
+                                isArray = true;
+                            }
                             // deprecated as of 2.4-jenkins-4 but left here for a while until we are sure nobody uses this
                             className = j.getString("stapler-class");
                             LOGGER.log(FINE, "stapler-class is deprecated in favor of $class: {0}", className);
                         }
                         if(j.has("$class")) {
+                            if (j.optJSONArray("$class") != null) {
+                                isArray = true;
+                            }
                             className = j.getString("$class");
                         }
 
                         if (className != null) {
                             // sub-type is specified in JSON.
                             // note that this can come from malicious clients, so we need to make sure we don't have security issues.
 
+                            if (isArray) {
+                                throw new IllegalArgumentException(
+                                        "The frontend sent an unexpected list of classes ("
+                                                + className
+                                                + ") rather than an expected single class. See"
+                                                + " https://www.jenkins.io/doc/developer/views/table-to-div-migration/"
```

Comment:
I am pretty sure tables-to-divs is not the only case where this problem arises, though it may be the most common one recently.

Reply:
Oh definitely include the link. I just meant it might be appropriate to adjust the phrasing to explain that this is only one possible root cause.

Reply:
I will keep an eye out for other causes in the future and adjust the message then if needed.